### PR TITLE
C* Timestamps, Part 3: ETE Test

### PR DESCRIPTION
**Pull request by jeremyk-91**
**Goals (and why)**:
- Have an ete test that checks that SSTables are written in a sane way

**Implementation Description (bullets)**:
- The test writes a value, overwrites it, and then forces the old version of the value to be swept.
It then checks the relevant SSTable, ensuring that it has a minimum timestamp that is recent (greater than the write timestamp of the first value) and that it has droppable tombstones.
- The test is run as part of a Cassandra+TimeLock test, and also on Cassandra+Embedded.

**Concerns (what feedback would you like?)**:
- Are the expectations of this test too strong?
- Tests that Atlas starts properly and behaves in a sane way already happen on the other paths, so it should be fine to just test the two paths I did.
- I kind of abused awaitility - is this _too_ gross?

**Where should we start reviewing?**: CassandraTimestampsEteTest

**Priority (whenever / two weeks / yesterday)**: this week or so

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/3266)

### CODE REVIEW

File: atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraTimestampsEteTest.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+                .until(() -> {
+                        todoClient.runIterationOfTargetedSweep();
+                        return todoClient.doesNotExistBeforeTimestamp(id, timestamp);
+                });
+    }
+
+    private void assertMinimumTimestampIsAtLeast(String sstableMetadata, long expectedMinimum) {
+        Pattern pattern = Pattern.compile("Minimum timestamp: ([0-9]*)");
+        Matcher timestampMatcher = pattern.matcher(sstableMetadata);
+        assertThat(timestampMatcher.find()).isTrue();
```
Comment (gsheasby) [inline atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraTimestampsEteTest.java]: Does this give a friendly error message when it fails?

File: atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraTimestampsEteTest.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        todoClient.truncate();
+    }
+
+    @After
+    public void cleanup() {
+        todoClient.truncate();
+    }
+
+    @Test
+    public void timestampsInOneSSTableAreCurrent() throws IOException, InterruptedException {
```
Comment (gsheasby) [inline atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraTimestampsEteTest.java]: Consider adding a comment about the implications if this test fails. Out of curiosity, have you attempted to run it on a pre-3224 branch?

File: atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/cassandra/util/CassandraCommands.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+public class CassandraCommands {
+    public static void nodetoolFlush(String containerName) throws IOException, InterruptedException {
+        EteSetup.execCliCommand(containerName, "nodetool flush");
+    }
+
+    public static List<String> nodetoolGetSSTables(String containerName,
+            String keyspace,
+            TableReference tableRef,
+            byte[] rowKey) throws IOException, InterruptedException {
+        String query = String.format("nodetool getsstables %s %s__%s %s",
```
Comment (gsheasby) [inline atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/cassandra/util/CassandraCommands.java]: is there a utility that converts a tablespace to %s__%s? We must have done this conversion when creating the table.

File: atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraTimestampsEteTest.java (hunk 1 above)
Comment (jeremyk-91) [reply atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraTimestampsEteTest.java]: No (it gives `expected: <true> but was: <false>`) but I've used assertj's `as` to give a more friendly error message (`[SSTableMetadata output contains the minimum timestamp] expected: <true> but was: <false>`). I did the same for the other regex match as well.

File: atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraTimestampsEteTest.java (hunk 2 above)
Comment (jeremyk-91) [reply atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraTimestampsEteTest.java]: Added several comments including an additional assertion that's important. Actually the todo table is thorough (with regards to the general comment I wanted to do one for a conservative table that I thought maybe could be a part 4).

File: atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/cassandra/util/CassandraCommands.java (hunk 3 above)
Comment (jeremyk-91) [reply atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/cassandra/util/CassandraCommands.java]: Yep - `CassandraKeyValueServiceImpl.internalTableName(tableRef)` should do it, changed to use that instead.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceImpl.java [hunk 4]
```diff
@@ -332,6 +332,7 @@ private CassandraKeyValueServiceImpl(Logger log,
             CassandraMutationTimestampProvider mutationTimestampProvider) {
         super(AbstractKeyValueService.createFixedThreadPool("Atlas Cassandra KVS",
                 config.poolSize() * config.servers().size()));
+//        mutationTimestampProvider = CassandraMutationTimestampProviders.legacyModeForTestsOnly();
```
Comment (gsheasby) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceImpl.java]: oops

File: atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraTimestampsEteTest.java [hunk 5]
```diff
@@ -58,13 +58,14 @@ public void cleanup() {
     }
 
     @Test
-    public void timestampsInOneSSTableAreCurrent() throws IOException, InterruptedException {
+    public void timestampsForRangeTombstonesAreCurrentInThoroughTables() throws IOException, InterruptedException {
```
Comment (gsheasby) [inline atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/CassandraTimestampsEteTest.java]: better name 👍

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceImpl.java (hunk 4 above)
Comment (jeremyk-91) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceImpl.java]: oops!

Comment (jeremyk-91) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceImpl.java]: and that's how I tested that things wouldn't work in the old world 😅

**Comment by jeremyk-91** [comment]
Something has made the startup independence ETE tests unhappy - investigating.

**Pull request merged**